tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,763 @@
+			shape[i] = []int64{int64(t.Dim(i))}
+		}
+	}
+
+	return shape
+}
+
+// Dim returns the size of the specified dimension.
+func (t *Tensor) Dim(n int) int {
+	return int(TF_Dim(t.tensor, n))
+}
+
+// DataSize returns the size of the data in bytes contained in a tensor.
+func (t *Tensor) DataSize() int64 {
+	return TF_TensorByteSize(t.tensor)
+}
+
+// Data returns the data contained in a tensor as a slice of bytes.
+func (t *Tensor) Data() []byte {
+	length := t.DataSize()
+	return (*[1 << 40]byte)(unsafe.Pointer(TF_TensorData(t.tensor)))[:length:length]
+}
+
+// String string representation of a tensor.
+func (t *Tensor) String() string {
+	return fmt.Sprintf("%v: dims:%v size:%v", t.DataType(), t.NumDims(), t.DataSize())
+}
+
+// AsStr returns the content of the tensor as slice of strings if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DtString
+func (t *Tensor) AsStr() (res [][]byte, err error) {
+	if DtString != t.DataType() {
+		err = &ErrInvalidTensorType{
+			tensorType:   t.DataType(),
+			expectedType: DtString,
+		}
+		return
+	}
+
+	if t.StringVal != nil {
+		return t.StringVal, nil
+	}
+
+	resultBytes := []byte{}
+	inStr := false
+	for _, b := range t.Data() {
+		if inStr {
+			if b == cBellByte {
+				res = append(res, resultBytes)
+				resultBytes = []byte{}
+			} else {
+				resultBytes = append(resultBytes, byte(b))
+			}
+		} else {
+			// TODO: Must be any better way to parse the strings...
+			if b == cAckByte || b == cBellByte || b == cDc1 {
+				inStr = true
+			}
+		}
+	}
+	if len(resultBytes) > 0 {
+		res = append(res, resultBytes)
+	}
+	t.StringVal = res
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return
+}
+
+// AsFloat32 returns the content of the tensor as a slice of float32 if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DtFloat
+func (t *Tensor) AsFloat32() (res []float32, err error) {
+	if DtFloat != t.DataType() {
+		err = &ErrInvalidTensorType{
+			tensorType:   t.DataType(),
+			expectedType: DtFloat,
+		}
+		return
+	}
+
+	if t.FloatVal != nil {
+		return t.FloatVal, nil
+	}
+
+	data := t.Data()
+	res = make([]float32, len(data)/cBytesFloat32)
+	for i := range res {
+		res[i] = math.Float32frombits(binary.LittleEndian.Uint32(data[i*cBytesFloat32 : (i+1)*cBytesFloat32]))
+	}
+	t.FloatVal = res
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return
+}
+
+// AsFloat64 returns the content of the tensor as a slice of float64 if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DtDouble
+func (t *Tensor) AsFloat64() (res []float64, err error) {
+	if DtDouble != t.DataType() {
+		err = &ErrInvalidTensorType{
+			tensorType:   t.DataType(),
+			expectedType: DtDouble,
+		}
+		return
+	}
+
+	if t.DoubleVal != nil {
+		return t.DoubleVal, nil
+	}
+
+	data := t.Data()
+	res = make([]float64, len(data)/cBytesFloat64)
```

Comment:
I love range, but I think your intent would read more clearly if you did:
numElements := len(data) / cBytesFloat64
res = make([]float64, numElements)
for i := 0; i < numElements; i++ { ... }

to make it obvious that it's filling res, as opposed to actually using something already in res.